GoogleSignIn/Tests/Unit/GIDSignInTest.m

```diff
@@ -503,6 +503,46 @@ - (void)testRestorePreviousSignInWhenSignedOut {
   [_authState verify];
 }
 
+- (void)testNotRestorePreviousSignInWhenSignedOutAndCompletionIsNil {
+  [[[_authorization expect] andReturn:_authState] authState];
+  [[[_authState expect] andReturnValue:[NSNumber numberWithBool:NO]] isAuthorized];
+
+  [_signIn restorePreviousSignInWithCompletion:nil];
+
+  XCTAssertNil(_signIn.currentUser);
+}
+
+- (void)testRestorePreviousSignInWhenCompletionIsNil {
+  [[[_authorization expect] andReturn:_authState] authState];
+  [[[_authState expect] andReturnValue:[NSNumber numberWithBool:YES]] isAuthorized];
+
+  OIDTokenResponse *tokenResponse =
+      [OIDTokenResponse testInstanceWithIDToken:[OIDTokenResponse fatIDToken]
+                                    accessToken:kAccessToken
+                                      expiresIn:nil
+                                   refreshToken:kRefreshToken
+                                   tokenRequest:nil];
+
+  [[[_authState stub] andReturn:tokenResponse] lastTokenResponse];
+
+  id profile = OCMStrictClassMock([GIDProfileData class]);
+  OCMStub([profile email]).andReturn(kUserEmail);
+  OCMStub([_user profile]).andReturn(profile);
+  OCMStub([_user userID]).andReturn(kUserID);
+
+  [[[_user stub] andReturn:_user] alloc];
+  (void)[[[_user expect] andReturn:_user] initWithAuthState:OCMOCK_ANY
+                                                profileData:OCMOCK_ANY];
+  XCTAssertNil(_signIn.currentUser);
+
+  [_signIn restorePreviousSignInWithCompletion:nil];
+
+  XCTAssertNotNil(_signIn.currentUser);
+  XCTAssertEqualObjects(_signIn.currentUser.userID, kUserID);
```

Comment:
Testing equality with `kUserID` feels a little vague and makes it seem like `kUserID` is sort of magical. I can't read this test and say that I understand why `_signIn.currentUser.userID == kUserID`.

Is it possible to do something like:

```
__block OIDAuthState *capturedAuthState;
__block GIDProfileData *capturedProfileData;
(void)[[[_user expect] andReturn:_user] initWithAuthState:SAVE_TO_ARG_BLOCK(capturedAuthState)
                                                profileData:SAVE_TO_ARG_BLOCK(capturedProfileData)];
XCTAssertEqualObjects(_signIn.currentUser.authState, capturedAuthState);
XCTAssertEqualObjects(_signIn.currentUser.profileData, capturedProfileData);
```

If that's not easy or possible (probably due to mocking complexity), then let's talk about what individual asserts we can add here to give us confidence in the test. Additionally, you can add a comment with a link to an issue on this repo describing the problem, which you can use when working on removing all of these mocks in this test file.

Reply:
I did try the above way, but it throw the exception:
testRestorePreviousSignInWhenCompletionIsNil: ((_signIn.currentUser.profile) equal to (capturedProfileData)) failed: throwing "NSInternalInconsistencyException: OCClassMockObject(GIDGoogleUser): unexpected method invoked: profile 
	stubbed:	alloc"

Reply:
Created the issue #306 to track dropping the mock user, will be implemented in the future.